### PR TITLE
fix: prevent stale-event notifications and inject crawl date into all LLM prompts

### LINKED ISSUE #497: Stale-event push notifications sent for past events (two compounding bugs)
## Summary

On **18 May 2026** a user received a push notification for message `JeZWVeHB` describing a football security operation at Vasil Levski stadium **on 16 May 2026** — two days after the event had already passed. Two independent bugs compounded to produce this outcome.

---

## Observed data

| Field | Value |
|---|---|
| `messageId` | `JeZWVeHB` |
| `source` | `sdvr-mvr-bg` |
| `crawledAt` | 2026-05-18 10:01:23 UTC+3 |
| `timespanStart` | 2026-05-18 10:01:23 UTC+3 (**= crawledAt fallback**) |
| `timespanEnd` | 2026-05-18 10:01:23 UTC+3 (**= crawledAt fallback**) |
| notification `matchedAt` | 2026-05-18 11:00:55 UTC+3 |

Source article: published on [mvr.bg on 15 May 2026](https://www.mvr.bg/sdvr/%D0%B8%D0%BD%D1%84%D0%BE%D1%80%D0%BC%D0%B0%D1%86%D0%B8%D0%BE%D0%BD%D0%B5%D0%BD-%D1%86%D0%B5%D0%BD%D1%82%D1%8A%D1%80/%D0%BF%D1%80%D0%B5%D1%81%D1%86%D0%B5%D0%BD%D1%82%D1%8A%D1%80/%D0%BD%D0%BE%D0%B2%D0%B8%D0%BD%D0%B8/90600).

Message text excerpt:
> "На **16 май** от **16.00 часа** на Националния стадион „Васил Левски" в София ще се проведе футболната среща…"

`process.extractLocations` audit stored on the message:
```json
{
  "step": "extractLocations",
  "summary": { "pinsCount": 1, "cityWide": true, "success": true },
  "timestamp": "2026-05-18T07:31:26.556Z"
}
```

Stored `pins[0]` from Firestore (the raw LLM output before `validateTimespanRange`):
```json
{
  "address": "Национален стадион „Васил Левски"",
  "timespans": [{ "start": "16.05.2024 14:00", "end": null }]
}
```

---

## Root Cause 1 — LLM guessed the wrong year (2024) for a year-less date

**File:** `ingest/prompts/extract-locations.md`  
**File:** `ingest/lib/ai-service.ts` → `extractLocations()`

The message says `"На 16 май от 16.00 часа"` — only a day and month, no year. The prompt gave the LLM no temporal context (no current date, no crawl date), so it guessed **2024**.

`validateTimespanRange` in `ingest/lib/timespan-utils.ts` correctly rejected `16.05.2024` as before the 2025-01-01 threshold and fell back to `crawledAt` = **18 May 2026**. The stored `timespanStart`/`timespanEnd` are both the crawl timestamp, making the message appear current.

---

## Root Cause 2 — No staleness guard in the notification pipeline

**File:** `ingest/notifications/message-fetcher.ts` → `getUnprocessedMessages()`

The only filter applied when fetching candidates for notification is `notificationsSent != true`. There is no check that `timespanEnd >= now`. Every un-notified message is eligible regardless of whether its event has already passed.

This is the missing defense-in-depth layer. Even if RC1 were fixed and the correct past date had been stored, the notification would still have fired.

---

## Reproduction path

1. Source publishes an article on **T-3** mentioning an event on **T-2** using only "DD Month" (no year).
2. Crawler scrapes the article on **T** (after the event has passed).
3. LLM guesses wrong year → `validateTimespanRange` rejects it → `timespanStart = timespanEnd = crawledAt = T`.
4. Notification job runs → message passes all filters → push notification sent for an event that ended at **T-2**.

### PR DESCRIPTION
**Pull request by vbuch**
Fixes #497 — two compounding bugs caused a push notification to be sent on 18 May 2026 for a football event that had already taken place on 16 May 2026.

## Root Cause 1 — LLM guesses wrong year for year-less dates

The source article (mvr.bg, 15 May) wrote `"на 16 май от 16.00 часа"` — day and month only, no year. The LLM guessed **2024**, which was rejected by `validateTimespanRange` (< 2025 threshold), so `timespanStart`/`timespanEnd` fell back to `crawledAt` (18 May 2026), making the message appear current.

### Fix

Extended `applyLocalityContext()` to accept optional `extraSubstitutions`. Added a `PromptContext` interface (`currentDate`, `sourceType`, `sourceUrl`) and injected `Today is {{CURRENT_DATE}}. Source: {{SOURCE_NAME}} ({{SOURCE_URL}}).` into all four LLM system prompts. The `extract-locations` year-inference rule now explicitly references `CURRENT_DATE`. `processWithAIPipeline` builds a `PromptContext` from `crawledAt` + source metadata and threads it to every AI call.

## Root Cause 2 — No staleness guard in the notification pipeline

`getUnprocessedMessages()` only filtered on `notificationsSent != true`. Any un-notified message was eligible regardless of whether its event had already ended.

### Fix

Map `timespanEnd` from Firestore and filter out messages where `timespanEnd < now`, logging each skip. Logic extracted as `isMessageStale()` — a pure exported function — for testability.

### Core fixes
- `ingest/lib/locality-context.ts` — `applyLocalityContext` accepts `extraSubstitutions?: Record<string, string>`
- `ingest/lib/ai-prompts.ts` — `PromptContext` type; `buildPromptExtras`; raw-file cache; all 4 prompt getters accept `ctx`
- `ingest/lib/ai-service.ts` — all 4 AI functions accept optional `ctx?: PromptContext`
- `ingest/messageIngest/index.ts` — `processWithAIPipeline` builds and threads `PromptContext`
- `ingest/notifications/message-fetcher.ts` — `isMessageStale()` + staleness filter in `getUnprocessedMessages`

### Prompt files
All four prompts (`filter-split.md`, `categorize.md`, `extract-locations.md`, `summarize.md`) now open with the date/source context line. The year-inference rule in `extract-locations.md` references `{{CURRENT_DATE}}` directly.

### Tests
- `ingest/lib/ai-prompts.test.ts` — `buildPromptExtras` date formatting and defaults
- `ingest/notifications/message-fetcher.test.ts` — `isMessageStale` boundary cases
- `ingest/lib/locality-context.test.ts` — `extraSubstitutions` merging and override behaviour

### Promptfoo evals
- `provider.ts` — reads `currentDate`/`sourceType`/`sourceUrl` from test vars and passes them as `PromptContext`
- `assertions.ts` — `assertTimespanYear` regression guard
- `extract-locations.yaml` — new test case with `currentDate: 2026-05-18` + `assertTimespanYear: 2026`
- `__mocks__/fixtures/sources/event-day-month-only.md` — fixture mirroring the production incident

### Documentation
- `e2e/message-ingest-and-notify.feature` — documents ingestion + notification scenarios

### CODE REVIEW

File: ingest/lib/ai-prompts.ts [hunk 1]
```diff
@@ -5,61 +5,114 @@ import { applyLocalityContext } from "@/lib/locality-context";
...
+  currentDate: Date;
+  /** Crawler / source identifier (e.g. "sdvr-mvr-bg"). */
+  sourceType?: string;
+  /** Base URL of the source (e.g. "https://www.mvr.bg/sdvr/…"). */
+  sourceUrl?: string;
+}
+
+/** Fallback context used when no crawl metadata is available (e.g. CLI tools). */
+const DEFAULT_PROMPT_CONTEXT: PromptContext = { currentDate: new Date() };
+
```
[the remark this answers is not in the document]
Comment (vbuch) [reply ingest/lib/ai-prompts.ts:23]: That should be rather fine. Since crawling may have been done hours earlier and the current date (as long as it is not spanning over multiple days) should be fine.

On a side note it may make sense for the `datePublished` to also be added to the context. But this may be done as part of https://github.com/oboapp/oboapp/issues/499 separately from the current change.
